### ISSUE TITLE
Agent: all scrapes in error after exactly 2 hours `msg="Appending scrape report failed" err="not found"`

### ISSUE THREAD

**Issue by fatpat** [prometheus/prometheus#9875]
**What did you do?**
I configured a prometheus in `Agent` mode. 

It was previously working in standalone mode and was working perfectly with the same scrapes configured.

**What did you expect to see?**
not a total shutdown of all scrapes after 2 hours

**What did you see instead? Under which circumstances?**
all scrapes in error with `component="scrape manager" scrape_pool=xxxx target="http://xxxx:6161/stat?format=prometheus" msg="Appending scrape report failed" err="not found"`

all exporters are reported down in the interface as well

After a restart everything is fine

**Environment**

* System information:
`Linux 5.11.0-27-generic x86_64`

* Prometheus version:
```
prometheus, version 2.32.0-beta.0 (branch: HEAD, revision: c32725ba7873dbaa39c223410043430ffa5a26c0)
  build user:       root@da630543d231
  build date:       20211116-11:23:14
  go version:       go1.17.3
  platform:         linux/amd64
```
* Prometheus configuration file:
```
global:
  scrape_interval: 15s
  scrape_timeout: 10s
  evaluation_interval: 15s
  external_labels:
    environment: test
scrape_configs:
- job_name: beanstalkd-exporter
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  metric_relabel_configs:
  - source_labels: [__name__]
    separator: ;
    regex: ^(?:beanstalkd_)?(.*)$
    target_label: __name__
    replacement: beanstalkd_$1
    action: replace
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/beanstalkd-exporter.yml
    refresh_interval: 5m
- job_name: memcached-exporter
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/memcached-exporter.yml
    refresh_interval: 5m
- job_name: bbbb
  honor_timestamps: true
  params:
    format:
    - prometheus
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /v3.0/forward/stats
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/bbbb.yml
    refresh_interval: 5m
- job_name: ttttt
  honor_timestamps: true
  params:
    format:
    - prometheus
  scrape_interval: 15s
  scrape_timeout: 10s
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/tttt.yml
    refresh_interval: 5m
- job_name: rrrr
  honor_timestamps: true
  params:
    format:
    - prometheus
  scrape_interval: 15s
  scrape_timeout: 10s
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/rrrr.yml
    refresh_interval: 5m
- job_name: netdata
  honor_timestamps: true
  params:
    format:
    - prometheus_all_hosts
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /api/v1/allmetrics
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/netdata.yml
    refresh_interval: 5m
- job_name: dddd
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/dddd.yml
    refresh_interval: 5m
- job_name: ccc
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/ccc.yml
    refresh_interval: 5m
- job_name: prometheus
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/prometheus.yml
    refresh_interval: 5m
- job_name: bbb
  honor_timestamps: true
  params:
    format:
    - prometheus
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /stat
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/bbb.yml
    refresh_interval: 5m
- job_name: aaa
  honor_timestamps: true
  params:
    format:
    - prometheus
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /status
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/aaaa.yml
    refresh_interval: 5m
- job_name: redis-exporter
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/redis-exporter.yml
    refresh_interval: 5m
- job_name: redissentinel-exporter
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  file_sd_configs:
  - files:
    - /etc/prometheus/targets/redissentinel-exporter.yml
    refresh_interval: 5m
remote_write:
- url: http://xxxxxxxx:9090/api/v1/write
  remote_timeout: 30s
  follow_redirects: true
  queue_config:
    capacity: 2500
    max_shards: 200
    min_shards: 1
    max_samples_per_send: 500
    batch_send_deadline: 5s
    min_backoff: 30ms
    max_backoff: 5s
  metadata_config:
    send: true
    send_interval: 1m
    max_samples_per_send: 500
```

* Logs:
```
Nov 26 17:03:10 prometheus[36155]: ts=2021-11-26T17:03:10.957Z caller=scrape.go:1258 level=warn component="scrape manager" scrape_pool=rawx target="http://1.1.1.1:6161/stat?format=prometheus" msg="Appending scrape report failed" err="not found"
Nov 26 17:03:11 prometheus[36155]: ts=2021-11-26T17:03:11.027Z caller=scrape.go:1258 level=warn component="scrape manager" scrape_pool=netdata target="http://1.1.1.1:1505/api/v1/allmetrics?format=prometheus_all_hosts" msg="Appending scrape report failed" err="not found"
```

**Comment by LeviHarrison** [comment]
Seems to be an issue with a series reference retrieved by hash not corresponding to a series.

https://github.com/prometheus/prometheus/blob/e63b85be4fd7a6425dc697adc8244065028dfae2/tsdb/agent/db.go#L681-L685

https://github.com/prometheus/prometheus/blob/e63b85be4fd7a6425dc697adc8244065028dfae2/tsdb/agent/db.go#L720-L723

cc @rfratto

**Comment by rfratto** [comment]
The "two hour" comment makes me feel like there's something wrong with the GC that's being missed by testing. It might be deleting series that are still active. 

#9964 will at least fix the issue here (because it doesn't return ErrNotFound anymore if a ref ID is invalid)

**Comment by rfratto** [comment]
(Sorry, closed by accident with a keyboard shortcut I didn't know existed 🤷)

#9964 fixes part of the problem here, since it won't return an ErrNotFound anymore if a ref ID is invalid. 

I think there's still another issue though. With a normal scrape interval of <4 minutes, no ref ID from the scraper should ever be stale. With the two hour comment, it seems like there's a bug in the GC code that isn't hit by testing yet, something like active series getting deleted when they shouldn't be.

**Comment by rfratto** [comment]
> it seems like there's a bug in the GC code that isn't hit by testing yet, something like active series getting deleted when they shouldn't be.

I can't find anything obvious in the code. We only delete a series when it its last write timestamp is lower than the lowest value for `prometheus_remote_storage_queue_highest_sent_timestamp_seconds` across all queues. 

The only thing I can think of right now is if there's a scrape target with a manually added timestamp that has clock skew from a different time zone. That could artificially increase `prometheus_remote_storage_queue_highest_sent_timestamp_seconds` and cause other series to get deleted prematurely.

**Comment by fatpat** [comment]
I have 0 metrics that are scraped with a timestamp.

I will try to identify if a particular scrape would cause this.
I will also try to use your PR


By the way, in fact all targets failed after 2 hours except one scrape which does not answer with a valid prometheus format. I have the following errors:
```
Nov 26 20:11:17 prometheus[50367]: ts=2021-11-26T20:11:17.253Z caller=scrape.go:1258 level=warn component="scrape manager" scrape_pool=xxx target="http://1.1.1.1:6122/status?format=prometheus" msg="Appending scrape report failed" err="not found"
```

This scrape does not rise the "appending scape report failed" error. It seems logic to me but it can maybe help for debugging

**Comment by fatpat** [comment]
Note: I opened another issue with the same cluster and server/configuration. I don't know if it could be related somehow: https://github.com/prometheus/prometheus/issues/9867

**Comment by fatpat** [comment]
I compiled prom from your PR. It's been more than 2 hours and the GC came 2 hours after start without any problems:
`Nov 27 08:29:59 prometheus-PR9664[156406]: ts=2021-11-27T08:29:59.756Z caller=db.go:557 level=info msg="series GC completed" duration=19.870029ms`

I'll keep monitoring stuff around to ensure PR definitely fix the issue.

**Comment by rfratto** [comment]
@fatpat Thanks for testing it out! Just curious, how many series are you seeing marked for deletion when a GC runs compared to the overall number of series? `prometheus_agent_deleted_series` tracks series that have been deleted from memory but not the WAL (yet), alongside `prometheus_agent_active_series` for series that exist both in memory and the WAL.

**Comment by fatpat** [comment]
![image](https://user-images.githubusercontent.com/822436/143740836-15edef1a-a774-4f4a-8d5d-d2f58616920e.png)
`prometheus_agent_deleted_series`: green
`prometheus_agent_active_series`: yellow

**Comment by fatpat** [comment]
fixed with #9664, closing
